### PR TITLE
[breaking] Add fast and full default configs

### PR DESCRIPTION
**Pull request by sebastianbenz**
Add two new standard configurations optimizing for either performance or performance optimization. This also improve the minimal bundle install size as all dependencies required by non essential transformation steps are now optional.

### CODE REVIEW

**Comment by patrickkettner** [comment]
Exciting!

File: packages/optimizer/README.md [hunk 1]
```diff
@@ -23,20 +23,36 @@ The performance optimizations can improve page rendering times by up to 50%. You
 
 ## Usage
```
Comment (sebastianbenz) [inline packages/optimizer/README.md:24]: @patrickkettner I've tried to document the two new modes. Wdyt? 

Fast and Full are placeholders...

**Comment by sebastianbenz** [comment]
@patrickkettner I think this is ready for a first review

File: packages/optimizer/README.md [hunk 2]
```diff
@@ -23,20 +23,36 @@ The performance optimizations can improve page rendering times by up to 50%. You
...
 
-### API
+AMP Optimizer provides two different default modes:
+
+1. **Fast:** this will perform only critical optimizations minimizing bundle install size and execution time. Use this when running AMP Optimizer in your critical rendering path, for example, when running AMP Optimizer in your server.
+2. **Full:** this will perform all available optimizations, resulting in the best AMP performance. This mode offers a better AMP developer experience as missing AMP specific markup and imports are added automaticallye. This maskes it a great choice when integrating it into a static site generator. **Important:** use only when running AMP Optimizer does not negatively affect the rendering time of your page, for example, when building a static site.
```
Comment (patrickkettner) [inline packages/optimizer/README.md:29]: nits

*automatically
*makes

File: packages/optimizer/README.md [hunk 3]
```diff
@@ -49,7 +65,56 @@ ampOptimizer.transformHtml(originalHtml).then((optimizedHtml) => {
...
+
+### Using `full` mode
+
+Install via:
+
+```
+npm install @ampproject/toolbox-optimizer terser postcss postcss-safe-parser cssnano-simple
+```
+
+Note: `full` mode requires additional dependencies.
```
Comment (patrickkettner) [inline packages/optimizer/README.md:97]: what happens if you don't install them?

File: packages/optimizer/README.md [hunk 4]
```diff
@@ -106,11 +163,11 @@ Automatically inject any missing markup required by AMP.
 
 #### `autoExtensionImport`
 
-Automatically import any missing AMP Extensions (e.g. amp-carousel).
+Automatically import any missing AMP Extensions (e.g. amp-carousel). This is not recommended if AMP Optimizer is run in the critical serving path.
```
Comment (patrickkettner) [inline packages/optimizer/README.md:166]: Have you noted all slow options? Would it make sense to be more formal/consistent about how they are documented? A flag of some kind, maybe?

File: packages/optimizer/README.md [hunk 5]
```diff
@@ -216,14 +273,13 @@ get an intrinsic layout. For image detection to work, an optional dependency
 
 #### `minify`
 
-Minifies the generated HTML output and inlined CSS.
+Minifies the generated HTML output. Requires the following `terser` to be installed when inlined `amp-script` should be minified. Combine with [separateKeyframes](#separateKeyframes) to also minify CSS.
```
Comment (patrickkettner) [inline packages/optimizer/README.md:276]: > <pre>Requires the following `terser` </pre>

should be

<pre>Requires [`terser`](https://www.npmjs.com/package/terser)</pre>

File: packages/optimizer/README.md [hunk 6]
```diff
@@ -216,14 +273,13 @@ get an intrinsic layout. For image detection to work, an optional dependency
...
 
 - name: `minify`
 - valid options: `[true|false]`
-- default: `true`
+- default: `false`
 - used by: [MinifyHtml](lib/transformers/MinifyHtml.js), [SeparateKeyframes](lib/transformers/SeparateKeyframes.js)
-
-**Warning:** this setting is not recommended when running AMP Optimizer in your backend on every request as execution time can increase by up to 7x.
+- additional dependency (when using amp-script): `npm install terser`
```
Comment (patrickkettner) [inline packages/optimizer/README.md:282]: why are they not included? to keep the bundle size down? May make sense to mention that at some point, somewhere

File: packages/optimizer/index.js [hunk 7]
```diff
@@ -19,16 +19,29 @@ const {
...
+  createFullOptimizer: (config = {}) => new DomTransformer(Object.assign({}, CONFIG_BUILD, config)),
+  createFastOptimizer: (config = {}) =>
+    new DomTransformer(Object.assign({}, CONFIG_RUNTIME, config)),
   TRANSFORMATIONS_AMP_FIRST,
-  TRANSFORMATIONS_MINIMAL,
   TRANSFORMATIONS_PAIRED_AMP,
+  CONFIG_BUILD,
+  CONFIG_RUNTIME,
+  /*
+   * @deprecated
```
Comment (patrickkettner) [inline packages/optimizer/index.js:43]: is this deprecation logged to the console somewhere?

File: packages/optimizer/lib/transformers/MinifyHtml.js [hunk 8]
```diff
@@ -131,7 +131,14 @@ class MinifyHtml {
   }
 
   async minifyAmpScript(child) {
+    if (!isDependencyInstalled('terser')) {
+      this.log_.warn(
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/MinifyHtml.js:135]: this is probably splitting hairs, but would `log_.error` be more appropriate? It seems this should be more severe than a warning

File: packages/optimizer/lib/transformers/OptimizeAmpBind.js [hunk 9]
```diff
@@ -19,19 +19,14 @@ const {skipNodeAndChildren} = require('../HtmlDomHelper');
 const {isTemplate} = require('../AmpConstants');
 
 /**
+ * Inserts i-amphtml-binding attributes on any elements which have bindings to allow the runtime to detect amp-bimd elements more efficiently.
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/OptimizeAmpBind.js:22]: Could be brain fog, but this was really hard to parse for me. I think below reads a clearer. @CrystalOnScript may have thoughts on it?

allow the runtime to detect amp-bimd elements more efficiently by adding i-amphtml-binding attributes elements that have bindings

File: packages/optimizer/lib/transformers/SeparateKeyframes.js [hunk 10]
```diff
@@ -45,9 +42,28 @@ class SeparateKeyframes {
   constructor(config) {
     this.log_ = config.log.tag('SeparateKeyframes');
     this.minify = config.minify !== false;
+    this.enabled = !!config.separateKeyframes;
+    if (
+      this.enabled &&
+      (!isDependencyInstalled('postcss') ||
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/SeparateKeyframes.js:48]: purely from a DX standpoint, would it make sense to have `isDependencyInstalled` take an array?

```js
const areDependenciesInstalled = require('../isDependencyInstalled');


areDependenciesInstalled(['postcss','postcss-safe-parser', 'cssnano-simple'])
```

File: packages/optimizer/lib/transformers/SeparateKeyframes.js [hunk 11]
```diff
@@ -45,9 +42,28 @@ class SeparateKeyframes {
...
+        !isDependencyInstalled('postcss-safe-parser') ||
+        !isDependencyInstalled('cssnano-simple'))
+    ) {
+      this.enabled = false;
+      this.log_.debug(
```
Comment (patrickkettner) [inline packages/optimizer/lib/transformers/SeparateKeyframes.js:53]: You were using `log_.warn` before, is there a reason to use `debug` here instead?

**Review by patrickkettner — APPROVED** [review]
nits and questions, but nothing major

excited to see it!

File: packages/cli/package.json [hunk 12]
```diff
@@ -37,7 +37,7 @@
     "@ampproject/toolbox-update-cache": "^2.8.0-canary.15",
     "minimist": "1.2.5",
     "minimist-options": "4.1.0",
-    "node-fetch": "2.6.1"
+    "node-fetch": "^2.6.1"
```
[the remark this answers is not in the document]
Comment (sebastianbenz) [reply packages/cli/package.json]: no reason - fixed

File: packages/optimizer/README.md (hunk 2 above)
Comment (sebastianbenz) [reply packages/optimizer/README.md]: done

File: packages/optimizer/README.md (hunk 3 above)
Comment (sebastianbenz) [reply packages/optimizer/README.md]: added explanation

File: packages/optimizer/README.md (hunk 4 above)
Comment (sebastianbenz) [reply packages/optimizer/README.md:166]: The slow options are listed in the table above. Not sure it makes sense to further document these.

File: packages/optimizer/README.md (hunk 5 above)
Comment (sebastianbenz) [reply packages/optimizer/README.md]: fixed

File: packages/optimizer/README.md (hunk 6 above)
Comment (sebastianbenz) [reply packages/optimizer/README.md:282]: Above I'm saying `*Fast:** this will perform only critical optimizations minimizing bundle install size and execution time. `

File: packages/optimizer/lib/transformers/MinifyHtml.js (hunk 8 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/MinifyHtml.js]: Good point. Consistently printing an error now.

File: packages/optimizer/lib/transformers/OptimizeAmpBind.js (hunk 9 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/OptimizeAmpBind.js]: agree, rephrased

File: packages/optimizer/lib/transformers/SeparateKeyframes.js (hunk 11 above)
Comment (sebastianbenz) [reply packages/optimizer/lib/transformers/SeparateKeyframes.js]: fixed

**Comment by schlessera** [comment]
@sebastianbenz:
This seems to have broken some of the tests. For example, the spec tests targeting the `AutoExtensionImporter` are not importing any extensions anymore.
